include/GafferSceneUI/Private/Inspector.h

```diff
@@ -0,0 +1,238 @@
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef GAFFERSCENEUI_INSPECTOR_H
+#define GAFFERSCENEUI_INSPECTOR_H
+
+#include "GafferSceneUI/Export.h"
+
+#include "GafferScene/SceneAlgo.h"
+#include "GafferScene/ScenePlug.h"
+
+#include "Gaffer/EditScope.h"
+
+#include "IECore/RefCounted.h"
+
+#include "boost/variant.hpp"
+
+namespace GafferSceneUI
+{
+
+namespace Private
+{
+
+/// Inspectors provide an abstraction for querying properties of a scene, and
+/// optionally making node graph edits to change those properties. They allow a
+/// small set of UI components to expose editable properties without needing to
+/// know their underlying details.
+///
+/// Inspectors are responsible for _what_ is inspected, but the UI components
+/// are responsible for the Context in which inspection happens. For example, a
+/// ParameterInspector knows which parameter to inspect, but the UI provides the
+/// location to inspect via the current context.
+///
+/// Inspectors are "EditScope aware", allowing the user to create new edits using
+/// a target EditScope. One of the main contributions of the Inspector base class
+/// is to encode the rules for interaction with EditScopes.
+///
+/// Notes for future work :
+///
+/// - This is temporarily in a Private namespace so that we can iterate on it
+///   without concerns about ABI breakage. Intended to be public eventually.
+/// - We want to generalise Inspectors so they can be used with images as well
+///   as scenes. Beware any scene-centric design decisions.
+/// - We want to use a TransformInspector to replace much of TransformTool::Selection.
+///   This has additional requirements such as knowing the `transformSpace` that a node
+///   works in. We think this information can be stored in a dedicated TransformHistory
+///   class provided by SceneAlgo, avoiding any need to specialise Inspector::Result.
+class GAFFERSCENEUI_API Inspector : public IECore::RefCounted, public boost::signals::trackable
+{
+
+	public :
+
+		IE_CORE_DECLAREMEMBERPTR( Inspector );
+		IE_CORE_FORWARDDECLARE( Result );
+
+		/// The name of the property being inspected.
```

Comment:
It _isn't_ a label, in that the formatting (e.g. camelCase or snake_case to UI Case) is the responsibility of the UI. And it does impact what is inspected in as much as an inspector with a different name would be inspecting something else. The reason for the Inspector to provide it is so that a UI can label the result of the inspection without any further knowledge about how the inspector was constructed.

Not sure what would help most in terms of commenting this - any suggestions?

Reply:
As discussed via IM, I had kinda missed that the only way to set this is via protected constructor, but it might still be nice to have a comment on the constructor something like "subclasses should provide a camelCase name that will be formatted by UI code to identify this inspector when it appears in UIs"

Reply:
Took a pass at this in f9ab283 (squashed into the original commit).